File: 086_VerticalOrderTraversal.cpp
/*
Problem: Vertical Order Traversal
Description: Print vertical order traversal of a binary tree.

Approach: Use map with horizontal distance as key, perform level order traversal.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<vector<int>> verticalTraversal(TreeNode* root) {
    vector<vector<int>> result;
    if (!root) return result;
    
    map<int, map<int, multiset<int>>> nodes;
    queue<pair<TreeNode*, pair<int, int>>> q;
    q.push({root, {0, 0}});
    
    while (!q.empty()) {
        auto p = q.front();
        q.pop();
        TreeNode* node = p.first;
        int x = p.second.first, y = p.second.second;
        
        nodes[x][y].insert(node->val);
        
        if (node->left) q.push({node->left, {x - 1, y + 1}});
        if (node->right) q.push({node->right, {x + 1, y + 1}});
    }
    
    for (auto& p : nodes) {
        vector<int> col;
        for (auto& q : p.second) {
            col.insert(col.end(), q.second.begin(), q.second.end());
        }
        result.push_back(col);
    }
    
    return result;
}

int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    
    vector<vector<int>> result = verticalTraversal(root);
    for (auto& col : result) {
        for (int x : col) cout << x << " ";
        cout << endl;
    }
    
    return 0;
}

/*
Time Complexity: O(n log n)
Space Complexity: O(n)
*/

File: 086_VerticalOrderTraversal.py
"""
Problem: Vertical Order Traversal
Description: Print vertical order traversal of a binary tree.

Approach: Use dictionary with horizontal distance as key, perform level order traversal.
"""

from typing import List, Optional
from collections import deque, defaultdict

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def verticalTraversal(root: Optional[TreeNode]) -> List[List[int]]:
    if not root:
        return []
    
    nodes = defaultdict(lambda: defaultdict(list))
    queue = deque([(root, 0, 0)])
    
    while queue:
        node, x, y = queue.popleft()
        nodes[x][y].append(node.val)
        
        if node.left:
            queue.append((node.left, x - 1, y + 1))
        if node.right:
            queue.append((node.right, x + 1, y + 1))
    
    result = []
    for x in sorted(nodes.keys()):
        col = []
        for y in sorted(nodes[x].keys()):
            col.extend(sorted(nodes[x][y]))
        result.append(col)
    
    return result

def main():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    
    print(verticalTraversal(root))

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n log n)
Space Complexity: O(n)
"""
